apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -745,23 +735,65 @@ private async Task<IList<PooledSession>> CreatePooledSessionsAsync(int batchSize
+                            bool wasHealthy = Interlocked.Exchange(ref _lastUnhealthyBatchSize, success ? 0 : batchSize) == 0;
+                            writeHealthChangedLog = wasHealthy != success;
                         }
+
+                        // We requested batchSize sessions to be created, but in reality actualCreatedSession were created, which are now live.
+                        // Note that actualCreatedSessions might be 0 either because the call failed or because no sessions were created.
+                        Interlocked.Add(ref _liveOrRequestedSessionCount, -batchSize + actualCreatedSessions);
+
+                        Interlocked.Add(ref _inFlightSessionCreationCount, -batchSize);
+                    }
+                    if(writeHealthChangedLog)
+                    {
+                        Parent._logger.Info(() => $"Session pool for {_databaseName} is now {(success ? "healthy" : "unhealthy")}.");
                     }
+                }
+            }
 
-                    // We requested batchSize sessions to be created, but in reality actualCreatedSession were created, which are now live.
-                    // Note that actualCreatedSessions might be 0 either because the call failed or because no sessions were created.
-                    Interlocked.Add(ref _liveOrRequestedSessionCount, -batchSize + actualCreatedSessions);
+            private bool ShouldCreateMoreSessions(out int batchSize)
+            {
+                lock (_lock)
+                {
+                    if (Healthy)
+                    {
+                        int poolSize = _readWriteSessions.Count + _readOnlySessions.Count;
+
+                        // Determine how many more sessions to create.
+                        // We want to make sure that if all existing and new requests for session creation succeed:
+                        // - All queuing callers will have a session
+                        // - The pool will have at least as many entries as the options specify
+                        // However, we don't want to end up with more than the maximum number of active sessions
+                        // in the options.
+
+                        // In reality, the current pool size should be 0 if pendingAcquisitionCount is non-zero, and
+                        // vice versa. However, we don't strictly enforce this, and the maths works out more simply if
+                        // we don't worry about it.
+
+                        // How many more sessions do we need to create in order to get the pool to the minimum size, after satisfying
+                        // pending callers, assuming that all the in-flight session requests succeed, and there are no more requests?
+                        int newSessionsToSatisfyPool = (_pendingAcquisitions.Count + Options.MinimumPooledSessions) - (poolSize + InFlightSessionCreationCount);
```

Comment:
I do wonder whether we can use LiveOrRequestedSessionCount here, but I'm getting slightly lost in all the counters we have :(

Reply:
LiveOrRequestedSessionCount includes the _pendingAcquisitions ones if I remember correctly.
I do agree that we should be able to simplify this, and I was actually looking into doing it on this PR but decided against it to mitigate risks.
I'm adding a TODO on my TODO list to look into this further.